### ISSUE TITLE
dpdk commands can not run inside kube-ovn container

### ISSUE THREAD

**Issue by wavezhang** [alauda/kube-ovn#406]
I install Kube-OVN with OVS-DPDK following the guide https://github.com/alauda/kube-ovn/blob/master/docs/dpdk.md
and replace ovs images with kubeovn/kube-ovn-dpdk:19.11.2

But when I'm trying to debug dpdk, and find the following command can not work
```
[root@node01 dpdk-stable-19.11.2]# kubectl exec -it -n kube-ovn ovs-ovn-wmmgw bash
[root@node01 /]# dpdk-pdump 
Illegal instruction
[root@node01 /]# dpdk-proc-info 
Illegal instruction
[root@node01 /]# 

```

DPDK kernel modules compile and install  following http://docs.openvswitch.org/en/latest/intro/install/dpdk/#installing 
with dpdk version 19.11.2.
Kernel Version: 4.14.172
Host OS Version: CentOS Linux release 7.4.1708 (Core)

**Comment by oilbeater** [comment]
@garyloug can you have a look into this? The kube-ovn-dpdk:19.11.2 is built by the master dockerfile.

**Comment by garyloug** [comment]
Hi @wavezhang 
I think you might be taking a few unnecessary steps here. There is no need to compile DPDK on the host and there is no need to replace any kube-ovn images to make this work. DPDK and OVS have been compiled and built into the kube-ovn-dpdk:19.11 image. DPDK kernel modules are disabled here to keep the image portable. There are no requirements on the host except hugepages.

To install is just 2 steps:
```
wget https://raw.githubusercontent.com/alauda/kube-ovn/release-1.2/dist/images/install.sh
bash install.sh --with-dpdk=19.11 
```
This will install Kube-OVN and automatically replace the kernel-ovs image with the ovs-dpdk image.

```
# kubectl exec -it ovs-ovn-xps4b -n kube-system /bin/bash
# dpdk-procinfo
EAL: Detected 88 lcore(s)
EAL: Detected 2 NUMA nodes
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket_87535_50233f7bd713c6
EAL: Selected IOVA mode 'PA'
EAL: Probing VFIO support...
EAL: PCI device 0000:03:00.0 on NUMA socket 0
EAL:   probe driver: 8086:1528 net_ixgbe
EAL: PCI device 0000:03:00.1 on NUMA socket 0
EAL:   probe driver: 8086:1528 net_ixgbe
EAL: PCI device 0000:81:00.0 on NUMA socket 1
EAL:   probe driver: 8086:1572 net_i40e
EAL: PCI device 0000:81:00.1 on NUMA socket 1
EAL:   probe driver: 8086:1572 net_i40e
EAL: PCI device 0000:81:00.2 on NUMA socket 1
EAL:   probe driver: 8086:1572 net_i40e
EAL: PCI device 0000:81:00.3 on NUMA socket 1
EAL:   probe driver: 8086:1572 net_i40e
EAL: PCI device 0000:83:00.0 on NUMA socket 1
EAL:   probe driver: 8086:1572 net_i40e
EAL: PCI device 0000:83:00.1 on NUMA socket 1
EAL:   probe driver: 8086:1572 net_i40e
EAL: PCI device 0000:83:00.2 on NUMA socket 1
EAL:   probe driver: 8086:1572 net_i40e
EAL: PCI device 0000:83:00.3 on NUMA socket 1
EAL:   probe driver: 8086:1572 net_i40e
dpdk-procinfo [EAL options] -- -p PORTMASK
  -m to display DPDK memory zones, segments and TAILQ information
  -p PORTMASK: hexadecimal bitmask of ports to retrieve stats for
  --stats: to display port statistics, enabled by default
  --xstats: to display extended port statistics, disabled by default
  --metrics: to display derived metrics of the ports, disabled by default
  --xstats-name NAME: to display single xstat id by NAME
  --xstats-ids IDLIST: to display xstat values by id. The argument is comma-separated list of xstat ids to print out.
  --stats-reset: to reset port statistics
  --xstats-reset: to reset port extended statistics
  --collectd-format: to print statistics to STDOUT in expected by collectd format
  --host-id STRING: host id used to identify the system process is running on
  --show-port: to display ports information
  --show-tm: to display traffic manager information for ports
  --show-crypto: to display crypto information
  --show-ring[=name]: to display ring information
  --show-mempool[=name]: to display mempool information
  --iter-mempool=name: iterate mempool elements to display content
EAL: Error - exiting with code: 1
  Cause: No Ethernet ports - bye
```

On your kubectl exec command you use namespace “kube-ovn”. I think kube-ovn has been using the namespace “kube-system” for quite a few versions now. Is this an older version?

**Comment by wavezhang** [comment]
I replace kube-ovn images because it does not read ovs config from file
![image](https://user-images.githubusercontent.com/832911/86988925-eb64e000-c1cb-11ea-8c89-79dcf8fc63c1.png)

I change namespace back to “kube-ovn” for consistence

**Comment by wavezhang** [comment]
since dpdk-devel is installed，
https://github.com/alauda/kube-ovn/blob/352f1e81cc42dd5584bb8ad84509992d5e2dde13/dist/images/Dockerfile.dpdk1911#L18-L19

why should we still build dpdk from source？
https://github.com/alauda/kube-ovn/blob/352f1e81cc42dd5584bb8ad84509992d5e2dde13/dist/images/Dockerfile.dpdk1911#L27-L35

**Comment by garyloug** [comment]
OVS-DPDK needs DPDK to be installed from source. It needs a target to build against. It’s covered in the link you sent: http://docs.openvswitch.org/en/latest/intro/install/dpdk/#installing

This would be enough if we were to simply “make install” OVS-DPDK. However, since we are installing OVN from RPM, it means OVS RPMs must be installed as a prerequisite. We must build and install OVS-DPDK RPMs. Using RPMs also helps keep the image size small with multistage docker build.

To build OVS-DPDK RPMs “make rpm-fedora” we need the additional package dpdk-devel. This is a separate package to the DPDK package we installed earlier from source. As mentioned at the top of the Dockerfile, dpdk-devel was the main reason for basing the OVS-DPDK image on Fedora. CentOS does not (at least at the time) have dpdk-devel v19.11 in the dnf/yum repos. Fedora32 repos had it.
https://github.com/alauda/kube-ovn/blob/352f1e81cc42dd5584bb8ad84509992d5e2dde13/dist/images/Dockerfile.dpdk1911#L1-L3

**Comment by garyloug** [comment]
> I replace kube-ovn images because it does not read ovs config from file

I see, I just checked and I have the same problem when I download kube-ovn-dpdk:19.11 image. It's using an older start-ovs-dpdk.sh script that does not check the ovs config file.

@oilbeater when was the kube-ovn-dpdk:19.11 image built? The ovs config feature that @wavezhang is looking for was added by @ipatrykx a few days [after the 1.2 release](https://github.com/alauda/kube-ovn/commit/8121afd6037a3414b2ca1a0ff8f3f004d47695a7). If the image was not built since 1.2 then it explains why the image still has the older script.

Can the kube-ovn-dpdk:19.11 image be rebuilt to include the [newer start-ovs-dpdk](https://github.com/alauda/kube-ovn/blob/master/dist/images/start-ovs-dpdk.sh) script?

**Comment by garyloug** [comment]
@oilbeater @wavezhang I just realized there is a 19.11.2 image..

What is on this image? @oilbeater you say in the comment above that it uses master dockerfile, but master docker file is still using DPDK 19.11.1
https://github.com/alauda/kube-ovn/blob/352f1e81cc42dd5584bb8ad84509992d5e2dde13/dist/images/Dockerfile.dpdk1911#L7

Is the plan to create an image for every DPDK sub version?

**Comment by oilbeater** [comment]
> @oilbeater @wavezhang I just realized there is a 19.11.2 image..
> 
> What is on this image? @oilbeater you say in the comment above that it uses master dockerfile, but master docker file is still using DPDK 19.11.1
> https://github.com/alauda/kube-ovn/blob/352f1e81cc42dd5584bb8ad84509992d5e2dde13/dist/images/Dockerfile.dpdk1911#L7
> 
> Is the plan to create an image for every DPDK sub version?

19.11.2 is a temporary image I build with the ovs config feature. This tag just for temporary use.

**Comment by wavezhang** [comment]
> OVS-DPDK needs DPDK to be installed from source. It needs a target to build against. It’s covered in the link you sent: http://docs.openvswitch.org/en/latest/intro/install/dpdk/#installing
> 
> This would be enough if we were to simply “make install” OVS-DPDK. However, since we are installing OVN from RPM, it means OVS RPMs must be installed as a prerequisite. We must build and install OVS-DPDK RPMs. Using RPMs also helps keep the image size small with multistage docker build.
> 
> To build OVS-DPDK RPMs “make rpm-fedora” we need the additional package dpdk-devel. This is a separate package to the DPDK package we installed earlier from source. As mentioned at the top of the Dockerfile, dpdk-devel was the main reason for basing the OVS-DPDK image on Fedora. CentOS does not (at least at the time) have dpdk-devel v19.11 in the dnf/yum repos. Fedora32 repos had it.
> https://github.com/alauda/kube-ovn/blob/352f1e81cc42dd5584bb8ad84509992d5e2dde13/dist/images/Dockerfile.dpdk1911#L1-L3

I'm facing a problem that if ovs-dpdk pod restart, all vhost port is will not work. I'm trying to figure out if it has something to do with dpdk/ovs build.

And i find that packets from the mac address are changing form vhost port.
```
ovs-appctl dpif/dump-flows -m br-int

ufid:ecbb67c6-6425-4b25-8224-11d4d3c5ef9a skb_priority(0/0),skb_mark(0/0),ct_state(0/0),ct_zone(0/0),ct_mark(0/0),ct_label(0/0),recirc_id(0),dp_hash(0/0),in_port(944a3fd8c27c-net1),packet_type(ns=0,id=0),eth(src=ff:ff:2a:1d:e2:42/01:00:00:00:00:00,dst=00:00:ff:ff:ff:ff/00:00:00:00:00:00),eth_type(0/0xffff), packets:17, bytes:748, used:1.457s, actions:drop
```
the actual mac address should be `2a:1d:e2:42:01:04`

**Comment by garyloug** [comment]
> I'm facing a problem that if ovs-dpdk pod restart, all vhost port is will not work. I'm trying to figure out if it has something to do with dpdk/ovs build.

I think this is to be expected. From [here](http://docs.openvswitch.org/en/latest/topics/dpdk/vhost-user/):

> vHost User uses a client-server model. The server creates/manages/destroys the vHost User sockets, and the client connects to the server. Depending on which port type you use, **dpdkvhostuser** or **dpdkvhostuserclient**, a different configuration of the client-server model is used.
> 
> For vhost-user ports, Open vSwitch acts as the server and QEMU the client. This means if OVS dies, all VMs must be restarted. On the other hand, for vhost-user-client ports, OVS acts as the client and QEMU the server. This means OVS can die and be restarted without issue, and it is also possible to restart an instance itself. For this reason, vhost-user-client ports are the preferred type for all known use cases; the only limitation is that vhost-user client mode ports require QEMU version 2.7. Ports of type vhost-user are currently deprecated and will be removed in a future release.

Userspace CNI is requesting socket type **dpdkvhostuser**, so restarting OVS is expected to cause problems. I think Userspace CNI can be set to use type **dpdkvhostuserclient** (See [here](https://github.com/intel/userspace-cni-network-plugin/blob/673eeb0d731f9de26bad2de7eb751f2159c02c99/cniovs/ovsctrl.go#L24-L27)) but I've never used this. To use this I think you must change the [Network Attachment Definition](https://github.com/alauda/kube-ovn/blob/master/docs/dpdk.md#userspace-network-attachment-definition). You'll see there is a host vhost mode and a container vhost mode. I think you must swap server and client.

By the way, FYI @oilbeater @wavezhang, I'm currently exploring what is needed to take the required functionality from Userspace-CNI and include it in Kube-OVN. I plan to open an issue/feature to discuss it when I have some more detail for you.

**Comment by garyloug** [comment]
> 19.11.2 is a temporary image I build with the ovs config feature. This tag just for temporary use.

Ok, I see. No other changes from current master?
It's strange, I tried your 19.11.2 image and I see the "Illegal instruction" error too but no errors on 19.11. There are no changes to the dockerfile, only the new start-ovs-dpdk script..

**Comment by oilbeater** [comment]
> Ok, I see. No other changes from current master?
> It's strange, I tried your 19.11.2 image and I see the "Illegal instruction" error too but no errors on 19.11. There are no changes to the dockerfile, only the new start-ovs-dpdk script..

@garyloug 19.11.2 also with the image size reduction optimization.  The 19.11 image comes from your repo before I just retag and push it to kubeovn repo. I wonder if my local dev env has something different from yours. Can you build a image form the master and test if the "illegal instruction" still exists?

**Comment by wavezhang** [comment]
> > I'm facing a problem that if ovs-dpdk pod restart, all vhost port is will not work. I'm trying to figure out if it has something to do with dpdk/ovs build.
> 
> I think this is to be expected. From [here](http://docs.openvswitch.org/en/latest/topics/dpdk/vhost-user/):
> 
> > vHost User uses a client-server model. The server creates/manages/destroys the vHost User sockets, and the client connects to the server. Depending on which port type you use, **dpdkvhostuser** or **dpdkvhostuserclient**, a different configuration of the client-server model is used.
> > For vhost-user ports, Open vSwitch acts as the server and QEMU the client. This means if OVS dies, all VMs must be restarted. On the other hand, for vhost-user-client ports, OVS acts as the client and QEMU the server. This means OVS can die and be restarted without issue, and it is also possible to restart an instance itself. For this reason, vhost-user-client ports are the preferred type for all known use cases; the only limitation is that vhost-user client mode ports require QEMU version 2.7. Ports of type vhost-user are currently deprecated and will be removed in a future release.
> 
> Userspace CNI is requesting socket type **dpdkvhostuser**, so restarting OVS is expected to cause problems. I think Userspace CNI can be set to use type **dpdkvhostuserclient** (See [here](https://github.com/intel/userspace-cni-network-plugin/blob/673eeb0d731f9de26bad2de7eb751f2159c02c99/cniovs/ovsctrl.go#L24-L27)) but I've never used this. To use this I think you must change the [Network Attachment Definition](https://github.com/alauda/kube-ovn/blob/master/docs/dpdk.md#userspace-network-attachment-definition). You'll see there is a host vhost mode and a container vhost mode. I think you must swap server and client.
> 
> By the way, FYI @oilbeater @wavezhang, I'm currently exploring what is needed to take the required functionality from Userspace-CNI and include it in Kube-OVN. I plan to open an issue/feature to discuss it when I have some more detail for you.

I‘m using vhostuserclient mode’

**Comment by garyloug** [comment]
> @garyloug 19.11.2 also with the image size reduction optimization. The 19.11 image comes from your repo before I just retag and push it to kubeovn repo. I wonder if my local dev env has something different from yours. Can you build a image form the master and test if the "illegal instruction" still exists?

Ok this is starting to make sense. I was surprised to see image 19.11 was able to run apps like dpdk-pdump when I tried it. If 19.11 is still the "big image" then it makes sense.

As part of image reduction I switched from older make build to the newer meson/ninja build. I found it was easier to configure away some unneeded accessories, like these apps, in an effort to reduce the image size (I didn't expect anyone to log into the ovs-ovn pod to do debugging).

With make build I think some of these apps are included by default, but for meson I think we need to explicitly include these. [See here](https://github.com/DPDK/dpdk/blob/e0a439466b9d63e3a5cc063efbf08784300b54c4/app/meson.build#L8-L21). If you wish to use apps like pdump, proc-info, etc. then we need to included a list of -Dapps here:
https://github.com/alauda/kube-ovn/blob/352f1e81cc42dd5584bb8ad84509992d5e2dde13/dist/images/Dockerfile.dpdk1911#L32

This will increase the image size.
